examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2756 @@
+    std::vector<double> old_unfiltered_density_values(n_q_points);
+    std::vector<double> old_unfiltered_density_multiplier_values(n_q_points);
+    std::vector<double> filtered_unfiltered_density_values(n_q_points);
+    std::vector<double> filter_adjoint_unfiltered_density_multiplier_values(
+      n_q_points);
+
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        cell_matrix = 0;
+        cell_rhs    = 0;
+
+        cell->get_dof_indices(local_dof_indices);
+
+        fe_values.reinit(cell);
+
+        lambda.value_list(fe_values.get_quadrature_points(), lambda_values);
+        mu.value_list(fe_values.get_quadrature_points(), mu_values);
+
+        fe_values[densities].get_function_values(nonlinear_solution,
+                                                 old_density_values);
+        fe_values[displacements].get_function_values(nonlinear_solution,
+                                                     old_displacement_values);
+        fe_values[displacements].get_function_divergences(
+          nonlinear_solution, old_displacement_divs);
+        fe_values[displacements].get_function_symmetric_gradients(
+          nonlinear_solution, old_displacement_symmgrads);
+        fe_values[displacement_multipliers].get_function_values(
+          nonlinear_solution, old_displacement_multiplier_values);
+        fe_values[displacement_multipliers].get_function_divergences(
+          nonlinear_solution, old_displacement_multiplier_divs);
+        fe_values[displacement_multipliers].get_function_symmetric_gradients(
+          nonlinear_solution, old_displacement_multiplier_symmgrads);
+        fe_values[density_lower_slacks].get_function_values(
+          nonlinear_solution, old_lower_slack_values);
+        fe_values[density_lower_slack_multipliers].get_function_values(
+          nonlinear_solution, old_lower_slack_multiplier_values);
+        fe_values[density_upper_slacks].get_function_values(
+          nonlinear_solution, old_upper_slack_values);
+        fe_values[density_upper_slack_multipliers].get_function_values(
+          nonlinear_solution, old_upper_slack_multiplier_values);
+        fe_values[unfiltered_densities].get_function_values(
+          nonlinear_solution, old_unfiltered_density_values);
+        fe_values[unfiltered_density_multipliers].get_function_values(
+          nonlinear_solution, old_unfiltered_density_multiplier_values);
+        fe_values[unfiltered_densities].get_function_values(
+          filtered_unfiltered_density_solution,
+          filtered_unfiltered_density_values);
+        fe_values[unfiltered_density_multipliers].get_function_values(
+          filter_adjoint_unfiltered_density_multiplier_solution,
+          filter_adjoint_unfiltered_density_multiplier_values);
+
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
```

Comment:
Here, you could use iterators from `FEValues`, like `fe_values.quadrature_point_indices()`.

Reply:
Just found these! That's really nice. :)